firmata/client.go

```diff
@@ -0,0 +1,679 @@
+package firmata
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"periph.io/x/conn/v3/gpio"
+	"periph.io/x/conn/v3/i2c"
+	"periph.io/x/conn/v3/onewire"
+	"periph.io/x/conn/v3/pin"
+)
+
+// These max values are for data bytes as, within firmata, data is 7 bits long.
+const (
+	MaxUInt8  uint8  = (1<<8 - 1) >> (8 / 8)
+	MaxUInt16 uint16 = (1<<16 - 1) >> (16 / 8)
+	MaxUInt24 uint32 = (1<<24 - 1) >> (24 / 8)
+	MaxUInt32 uint32 = (1<<32 - 1) >> (32 / 8)
+)
+
+var commandResponseMap = map[SysExCmd]SysExCmd{
+	SysExAnalogMappingQuery: SysExAnalogMappingResponse,
+	SysExCapabilityQuery:    SysExCapabilityResponse,
+	SysExPinStateQuery:      SysExPinStateResponse,
+}
+
+type ClientI interface {
+	SendSysEx(SysExCmd, ...byte) (chan []byte, error)
+	SendReset() error
+	ExtendedReportAnalogPin(uint8, uint8) error
+	CapabilityQuery() (chan CapabilityResponse, error)
+	PinStateQuery(uint8) (chan PinStateResponse, error)
+	ReportFirmware() (chan FirmwareReport, error)
+	SetPinMode(uint8, pin.Func) error
+	SetAnalogPinReporting(uint8, bool) error
+	SetDigitalPinReporting(uint8, bool) error
+	SetDigitalPortReporting(uint8, bool) error
+	SetSamplingInterval(uint16) error
+	SetDigitalPinValue(p uint8, value gpio.Level) error
+	SendAnalogMappingQuery() (chan AnalogMappingResponse, error)
+	AnalogPinToDigitalPin(p uint8) (uint8, error)
+	SetAnalogIOMessageListener(p uint8, ch chan uint16) (release func(), err error)
+	SetDigitalIOMessageListener(p uint8, ch chan gpio.Level) (release func(), err error)
+	SendAnalogIOMessage(uint8, uint16) error
+
+	OpenI2CBus() (i2c.Bus, error)
+	SetI2CAddressListener(addr uint8, ch chan I2CPacket) (release func(), err error)
+	WriteI2CData(address uint8, restart bool, data []uint8) error
+	ReadI2CData(address uint8, restart bool, len uint16) error
+	ReadI2CRegister(address uint8, restart bool, register uint8, len uint16) error
+	SendI2CConfig(delayMicroseconds uint8) error
+
+	OpenOneWireBus(p uint8) (bus onewire.BusCloser, err error)
+	SetOneWireListener(uint8, chan []byte) (release func(), err error)
+
+	GetPinName(uint8) string
+	GetPinFunctions(uint8) []pin.Func
+
+	Close() error
+}
+
+type Client struct {
```

Comment:
Can you either export the interface or the struct but not both? I generally prefer the struct unless there's a clear reason like to make a fake/mock easier to use in practice.